### PR TITLE
Initial release

### PR DESCRIPTION
**Pull request by luigibk**
https://tools.hmcts.net/jira/browse/RPE-950

### CODE REVIEW

**Review by timja — DISMISSED** [review]
Separate code from initial setup please, much easier to review then

File: README.md
Comment (timja) [inline README.md]: isn't it about reading properties from the file system? I thought that was the point of not putting flexvolume in the name

File: README.md
Comment (timja) [inline README.md]: could you give an example from spring properties to file system?
i.e. 
```
draft-store.primary-encryption-key
```
would be:
```
/kvmnt/draft-store/primary-encryption-key
```

I think

File: README.md
Comment (timja) [inline README.md]: file in this repo is called `LICENSE` for some reason (no extension) so link  doesn't work

File: build.gradle
Comment (timja) [inline build.gradle]: 8.17 is the latest version

File: build.gradle
Comment (timja) [inline build.gradle]: 6.11.0

File: src/main/java/uk/gov/hmcts/reform/propertiesvolume/PropertiesvolumeConfigProperties.java
Comment (timja) [inline src/main/java/uk/gov/hmcts/reform/propertiesvolume/PropertiesvolumeConfigProperties.java:13]: This looks weird to  me, how about?
```suggestion
public class PropertiesVolumeConfigProperties {
```

File: build.gradle
Comment (timja) [inline build.gradle:9]: This plugin is reporting some updates that should be done:

```
$ ./gradlew dependencyUpdates

> Task :dependencyUpdates

------------------------------------------------------------
: Project Dependency Updates (report to plain text file)
------------------------------------------------------------

The following dependencies are using the latest milestone version:
 - com.github.ben-manes.versions:com.github.ben-manes.versions.gradle.plugin:0.20.0
 - com.jfrog.bintray:com.jfrog.bintray.gradle.plugin:1.8.4
 - io.spring.dependency-management:io.spring.dependency-management.gradle.plugin:1.0.6.RELEASE
 - net.ltgt.apt:net.ltgt.apt.gradle.plugin:0.21
 - org.junit.jupiter:junit-jupiter-api:5.4.0
 - org.junit.jupiter:junit-jupiter-engine:5.4.0
 - org.owasp.dependencycheck:org.owasp.dependencycheck.gradle.plugin:4.0.2
 - org.sonarqube:org.sonarqube.gradle.plugin:2.7
 - org.springframework.boot:spring-boot-configuration-processor:none
 - org.springframework.cloud:spring-cloud-context:2.1.0.RELEASE

The following dependencies have later milestone versions:
 - org.springframework.boot:org.springframework.boot.gradle.plugin [2.1.2.RELEASE -> 2.1.3.RELEASE]
 - org.springframework.boot:spring-boot-starter-test [2.1.2.RELEASE -> 2.1.3.RELEASE]
     https://projects.spring.io/spring-boot/#/spring-boot-parent/spring-boot-starters/spring-boot-starter-test
 - org.springframework.boot:spring-boot-starter-validation [2.1.2.RELEASE -> 2.1.3.RELEASE]
     https://projects.spring.io/spring-boot/#/spring-boot-parent/spring-boot-starters/spring-boot-starter-validation
 - org.springframework.boot:spring-boot-starter-web [2.1.2.RELEASE -> 2.1.3.RELEASE]
     https://projects.spring.io/spring-boot/#/spring-boot-parent/spring-boot-starters/spring-boot-starter-web
 - org.springframework.boot:spring-boot-starter-webflux [2.1.2.RELEASE -> 2.1.3.RELEASE]
     https://projects.spring.io/spring-boot/#/spring-boot-parent/spring-boot-starters/spring-boot-starter-webflux

Gradle updates:
 - Gradle: [5.2 -> 5.2.1]
```

File: src/main/java/uk/gov/hmcts/reform/propertiesvolume/PropertiesvolumeConfigProperties.java
Comment (timja) [inline src/main/java/uk/gov/hmcts/reform/propertiesvolume/PropertiesvolumeConfigProperties.java:23]: new line needed here?

File: src/main/java/uk/gov/hmcts/reform/propertiesvolume/PropertiesvolumeException.java
Comment (timja) [inline src/main/java/uk/gov/hmcts/reform/propertiesvolume/PropertiesvolumeException.java:3]: same here:
```suggestion
public class PropertiesVolumeException extends RuntimeException {
```

File: src/main/java/uk/gov/hmcts/reform/propertiesvolume/PropertiesvolumePropertySourceLocator.java
Comment (timja) [inline src/main/java/uk/gov/hmcts/reform/propertiesvolume/PropertiesvolumePropertySourceLocator.java:15]: thoughts on?
```suggestion
public class PropertiesVolumePropertySourceLocator implements PropertySourceLocator {
```

File: src/main/java/uk/gov/hmcts/reform/propertiesvolume/PropertiesvolumePropertySourceLocator.java
Comment (timja) [inline src/main/java/uk/gov/hmcts/reform/propertiesvolume/PropertiesvolumePropertySourceLocator.java:17]: didn't seem like the right order to me but looks good 👍 
https://stackoverflow.com/a/16732059

**Review by timja — COMMENTED** [review]
Code looks good, just some version updates and slight tweaks would be good 👍

File: src/main/java/uk/gov/hmcts/reform/propertiesvolume/PropertiesvolumeConfigProperties.java
Comment (luigibk) [reply src/main/java/uk/gov/hmcts/reform/propertiesvolume/PropertiesvolumeConfigProperties.java:13]: I prefer to keep it consistent with the way the property itself is named.

File: src/main/java/uk/gov/hmcts/reform/propertiesvolume/PropertiesvolumePropertySource.java
Comment (KevinJCross) [inline src/main/java/uk/gov/hmcts/reform/propertiesvolume/PropertiesvolumePropertySource.java:81]: hmmm ... should we really trim ?

File: src/main/java/uk/gov/hmcts/reform/propertiesvolume/PropertiesvolumePropertySource.java
Comment (KevinJCross) [inline src/main/java/uk/gov/hmcts/reform/propertiesvolume/PropertiesvolumePropertySource.java]: so we don't distinguish separate vaults ?
What about name-space clashes ?

File: src/main/java/uk/gov/hmcts/reform/propertiesvolume/PropertiesvolumePropertySource.java
Comment (luigibk) [reply src/main/java/uk/gov/hmcts/reform/propertiesvolume/PropertiesvolumePropertySource.java:81]: I think so. Also please do keep in mind that some choices (such as this) are consistent with spring-cloud-kubernetes-starter conventions.

File: src/main/java/uk/gov/hmcts/reform/propertiesvolume/PropertiesvolumePropertySource.java
Comment (luigibk) [reply src/main/java/uk/gov/hmcts/reform/propertiesvolume/PropertiesvolumePropertySource.java]: That is really a flexvolume limitation as overlapping mounts have a last one wins semantics.

**Comment by KevinJCross** [comment]
If we sort out the namespace conflicts im Ok to approve.
the nodejs does `secrets.vault.property-name` It would be nice to have the vault distinction in there :smile:

**Comment by KevinJCross** [comment]
Hmmm ... what ordering is this property source and how is it enforced ?
https://docs.spring.io/spring-boot/docs/current/reference/html/boot-features-external-config.html
Seems we need to be after the Random and before the profile.

File: src/main/java/uk/gov/hmcts/reform/propertiesvolume/PropertiesvolumePropertySource.java
Comment (KevinJCross) [reply src/main/java/uk/gov/hmcts/reform/propertiesvolume/PropertiesvolumePropertySource.java]: do we have the vault name in the property name ?
from my review its missing ?

File: src/main/java/uk/gov/hmcts/reform/propertiesvolume/PropertiesvolumePropertySource.java
Comment (KevinJCross) [reply src/main/java/uk/gov/hmcts/reform/propertiesvolume/PropertiesvolumePropertySource.java:81]: kewl ... np

File: src/test/java/uk/gov/hmcts/reform/propertiesvolume/PropertiesvolumesSpringBootTest.java
Comment (KevinJCross) [inline src/test/java/uk/gov/hmcts/reform/propertiesvolume/PropertiesvolumesSpringBootTest.java:37]: maybe we should have a test that has some value substitutions in an application.yaml?

Comment (KevinJCross) [reply src/test/java/uk/gov/hmcts/reform/propertiesvolume/PropertiesvolumesSpringBootTest.java:37]: this would prove the ordering of the property source.

File: checkstyle/checkstyle.xml
Comment (timja) [inline checkstyle/checkstyle.xml]: if you're going to move files like this do it in another PR please,
but all of our other repos have it under `config` to not clutter the root directory:
https://github.com/hmcts/draft-store/tree/master/config

File: README.md
Comment (timja) [inline README.md:54]: this differs to the nodejs implementation in that the keyvault name isn't prefixed,

I would have expected `draft-store` to be prefixed here.

Can the implementations be consistent please?

File: src/main/java/uk/gov/hmcts/reform/propertiesvolume/PropertiesvolumePropertySource.java
Comment (KevinJCross) [inline src/main/java/uk/gov/hmcts/reform/propertiesvolume/PropertiesvolumePropertySource.java:18]: looks like you need to add something like 
the addToEnvironment bit in the RandomPropertySource
https://github.com/spring-projects/spring-boot/blob/master/spring-boot-project/spring-boot/src/main/java/org/springframework/boot/env/RandomValuePropertySource.java

**Review by KevinJCross — DISMISSED** [review]
Please address these issues and Ill be happy to approve.
- Ordering of the property source (after Random) needs to be addressed and an application.yaml test to verify the ordering.
- Property namespace conflicts (inter vault or inter folder /cmc/prop1 and /s2s/prop1) need to be addressed via use of the vault/folder in the property name ( if not its not there already :smile:)

File: README.md
Comment (luigibk) [reply README.md:54]: Not necessarily users will mount flexvolumes with the keyvault name. What happens when the mount is: 

'/kvmnt/primary-encryption-key' or '/primary-encryption-key'?

File: checkstyle/checkstyle.xml
Comment (luigibk) [reply checkstyle/checkstyle.xml]: I did not move anything, I just merged your changes. By the way, can you please make a PR instead of making changes to this branch?

File: src/main/java/uk/gov/hmcts/reform/propertiesvolume/PropertiesvolumePropertySource.java
Comment (luigibk) [reply src/main/java/uk/gov/hmcts/reform/propertiesvolume/PropertiesvolumePropertySource.java:18]: Why would I need to do that?

File: src/main/java/uk/gov/hmcts/reform/propertiesvolume/PropertiesvolumePropertySource.java
Comment (luigibk) [reply src/main/java/uk/gov/hmcts/reform/propertiesvolume/PropertiesvolumePropertySource.java]: Not necessarily users will mount flexvolumes with the keyvault name. What happens when the mount is:

'/kvmnt/primary-encryption-key' or '/primary-encryption-key'?

File: checkstyle/checkstyle.xml
Comment (timja) [reply checkstyle/checkstyle.xml]: I could have done that but I just merged master to help you out...

Comment (timja) [reply checkstyle/checkstyle.xml]: i.e. you haven't merged properly after pulling my merge

File: src/test/java/uk/gov/hmcts/reform/propertiesvolume/PropertiesvolumesSpringBootTest.java
Comment (luigibk) [reply src/test/java/uk/gov/hmcts/reform/propertiesvolume/PropertiesvolumesSpringBootTest.java:37]: If you are referring to your comment above, I believe ordering is enforced by the PropertySourceLocator. Again I made this consistent with the implementations for Secrets and ConfigMap in spring-cloud-kubernetes.

File: src/main/java/uk/gov/hmcts/reform/propertiesvolume/PropertiesvolumeConfigProperties.java
Comment (timja) [reply src/main/java/uk/gov/hmcts/reform/propertiesvolume/PropertiesvolumeConfigProperties.java:13]: What do you meant by that? Its two words, shouldn't It be snakecased?

File: README.md
Comment (timja) [inline README.md:25]: would it be possible for you to link a PR here showing this being used on a java app?
I would like to see two things,
1. it working
2. if you aren't using flexvolume in the app it also works, so that we can merge code to master even though the target environment is still and App service, i.e. it works for AKS, local, and App services

Comment (luigibk) [reply README.md:25]: @timja: Sure. Which applications would you like to target for each type you mention?
@KevinJCross: can you please check if I interpreted your requests correctly?

Comment (timja) [reply README.md:25]: I had a look for one of ours, but I don't think we have any that have secrets that have got charts written,
this could be a good one:
https://github.com/hmcts/cmc-claim-store/blob/master/charts/cmc-claim-store/values.template.yaml
 
otherwise plum-recipes would be the other option

File: README.md
Comment (KevinJCross) [inline README.md]: hmmm ... do we really need enabled. Surely if this exists in the configuration it is enabled ?

File: README.md
Comment (luigibk) [reply README.md:25]: @timja isn't that only aks (i.e. helm chart)? what about local and app service?

File: README.md
Comment (KevinJCross) [inline README.md]: Why give the option. We should really just prefix or not since this is the initial implementation.
Is there a reason for adding a flag here or is this normally the default ?

File: README.md
Comment (timja) [reply README.md:25]: for app service you would have to deploy one of them to an app service (probably sandbox or saat) with the same code change as in the PR for the AKS one

Local is the same as app service really I think

File: README.md
Comment (luigibk) [reply README.md]: It's for consistency with spring-cloud-kubernetes to be honest. I am not very convinced by it either. It defaults to true.

**Review by timja — APPROVED** [review]
Code looks good,
I would like to see https://github.com/hmcts/properties-volume-spring-boot-starter/pull/1#discussion_r258938404 addresses but won't block on it.

Before releasing this the testing should be done to make sure this works without the files being present

File: README.md
Comment (KevinJCross) [inline README.md:23]: I personally would leave these out fo the example configuration since they are the defaults( I probably would only make them configurable if we actually have a use case too)
I say this because it confuses the user who should really never have to use these parameters.
Add a section after the example with the more detailed options so this is not copy pasted all over the place.

File: README.md
Comment (KevinJCross) [reply README.md:25]: I would suggest just putting just `secrets` into the paths parameter (simple easy and covers our use case) and let the code find the directories and use them as a prefix to fill out the properties.
i.e. from above `paths = 'secrets' `

This way the property names would be secrets.this.aProperty or `secrets.that.a-property` using the folder/file structure to name the property e.g secrets/that/a_prop would be secrets.that.a_prop. It should be simple to File.walk and represent the folder and file structure in a property :smile: and it makes the module configuration super simple.

From my perspective representing the properties like this is basically  like a folder/file representation of the yaml or json object structure and is a very versatile way of representing/reading the properties.

File: README.md
Comment (KevinJCross) [inline README.md]: could we not mount this in the /mnt/certs ?

File: README.md
Comment (KevinJCross) [inline README.md:54]: do we really need this option ?

File: src/test/java/uk/gov/hmcts/reform/propertiesvolume/PropertiesvolumesSpringBootTest.java
Comment (KevinJCross) [reply src/test/java/uk/gov/hmcts/reform/propertiesvolume/PropertiesvolumesSpringBootTest.java:37]: Could we have an application.yaml that uses the property substitution to check the ordering of the property sources ?
maybe something like this application.yaml
```yaml
someApplicationProperty:  "${secrets.kvcreds-test-single.kv5}"
```

**Pull request merged**
